### PR TITLE
View's queryPage doesn't seem to handle multi-value values

### PR DESCRIPTION
**Pull request by emlaver**
*What:*
View’s queryPage doesn’t seem to handle multi-value values (FB 48355)
The queryPage method doesn’t handle views that emit keys such as [“uuid”, 100].

#39 DB view key varargs needs parameter typing (FB 48532)
The key, start key, and end key methods are not able to handle complex keys.  One example is a complex key with multiple integer values i.e. key(2011 10, 15).

*How:*
The queryView method in View’s queryPage had a hard cast of String.  This would cause an exception if an integer value existed in a key from the result rows. To fix this:
The String parameter was changed to Object class to allow for other data types. 
If the key in the result rows is an array of multiple values (i.e. startkey=[“uuid”,100]), the array will be converted into one string and then be added as a property for the page parameters.

View’s query had a similar problem where multiple number value keys were not property handled when building the URI.  During testing, if a string and integer value existed in a key array, then the URI would be properly created. If the key only contained an array of numbers, then it would throw an exception. 
The fix required a statement in getKeyAsJson to check whether the multi value key array was exactly one value and an instance of the Number class.

*Testing:*
An existing test case under ViewsTest (byComplexKey) was modified to use the new multi value change.
New test cases were added to ViewsTest:
Assert that result list is not empty using multi value keys in both View’s query and queryPage method.
Run tests with multi value key arrays from one to three values

reviewer @ricellis 
reviewer @mikerhodes

### CODE REVIEW

File: src/test/java/com/cloudant/tests/ViewsTest.java [hunk 1]
```diff
@@ -594,6 +731,91 @@ private static void init() {
...
+            keyViews.put("created", mapFunctions);
+
+            designDoc.setViews(keyViews);
+
+            db.save(designDoc);
+
+            db.syncDesignDocsWithDb();
+
+        } catch (DocumentConflictException e) {
+        }
```
Comment (mikerhodes) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Conflicts should probably fail the test?

File: src/test/java/com/cloudant/tests/ViewsTest.java [hunk 2]
```diff
@@ -105,12 +108,146 @@ public void byStartAndEndKey() {
...
+
+
+        query = db.view("keyarray_example/creator_created_boolean")
+                .startKey(new Object[]{true, "uuid", 1})
+                .endKey(new Object[]{true, "uuid", 2000})
+                .includeDocs(true);
+
+        result = query.query(Object.class);
+
+        assertThat(result.size(), is(2));
```
Comment (mikerhodes) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: I assume a tear-down is cleaning up the database for you?

File: src/test/java/com/cloudant/tests/ViewsTest.java [hunk 3]
```diff
@@ -105,12 +108,146 @@ public void byStartAndEndKey() {
...
+            page = query.queryPage(30, null, Object.class);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        assertThat(page.getResultList().size(), is(3));
+    }
+
+    @Test
+    public void byStartAndEndThreeKeyArray() {
```
Comment (mikerhodes) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Add java doc for test purpose.

File: src/main/java/org/lightcouch/View.java [hunk 4]
```diff
@@ -373,7 +373,19 @@ public boolean queryForBoolean() {
...
+        Object getKey = rows.get(0).getKey();
+        //Handle multi value key array
+        if(getKey instanceof ArrayList) {
+            String keyAsString = "[";
+            for(Object key : (ArrayList<Object>)getKey) {
+                keyAsString += key.toString() + ",";
+            }
+            //Final key value as string with closing bracket
+            getKey = keyAsString.substring(0,keyAsString.length()-1).concat("]");
+        }
```
Comment (mikerhodes) [inline src/main/java/org/lightcouch/View.java]: Seems like you are trying to make a JSON array here. As far as I can see, this will make invalid JSON for anything other than numbers and boolean values, unless the keys are already pre-encoded as JSON (i.e., escaped and surrounded in quotes).

Why not use GSON here rather than manual concatenation?

**Comment by mikerhodes** [comment]
> The queryPage method doesn’t handle views that emit values such as [“uuid”, 100].

Do you mean "emit keys such as" here rather than "values"?

File: src/test/java/com/cloudant/tests/ViewsTest.java (hunk 2 above)
Comment (emlaver) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Yes, the tearDown method is deleting the whole test database after all tests are completed.

**Comment by emlaver** [comment]
Yes, I meant "keys".  Fixed.

File: src/main/java/org/lightcouch/View.java (hunk 4 above)
Comment (ricellis) [inline src/main/java/org/lightcouch/View.java]: Yeah it would be cleaner if it can be a JsonArray rather than trying to manipulate the string

**Comment by ricellis** [comment]
I think it might be worth having an additional test that uses a non-String, non-array key type to make sure that path works with these changes. Maybe something like `byStartAndEndIntKey()`

**Comment by emlaver** [comment]
@mikerhodes @ricellis I've added your comments in the latest commit.  I removed the manual concatenation code.  Now in getKeyAsJson, booleans and numbers are be passed as an object array to toJson.  With that, there's no extra code needed (just use toString) to convert the Object array to string for the page parameters.

**Comment by mikerhodes** [comment]
I suggest adding tests where pagination is required: it looks like the changes [here](https://github.com/cloudant/java-cloudant/pull/63/files#diff-462176273836fe6f0e80f32ba1f1aeadR386) affect pagination as they're in an `if` check for `!isLastPage`.

**Comment by emlaver** [comment]
@mikerhodes The message will be separated during squashing. I'll also add a test case for pagination.

Just FYI, I compared the debug test results in the console from master branch with this branch.  The values for paging parameters were the same.

File: src/main/java/org/lightcouch/View.java [hunk 5]
```diff
@@ -583,7 +585,14 @@ public View tempView(MapReduce mapReduce) {
...
     private String getKeyAsJson(Object... key) {
-        return (key.length == 1) ? gson.toJson(key[0]) : gson.toJson(key); // single or complex key
+        // single or complex key
+        if (key.length == 1 && (key[0] instanceof Boolean || key[0] instanceof Number)) {
+            return gson.toJson(key);
+        } else if (key.length == 1) {
+            return gson.toJson(key[0]);
+        } else {
+            return gson.toJson(key);
+        }
```
Comment (mikerhodes) [inline src/main/java/org/lightcouch/View.java]: Why does `[1]` and `[true]` get special treatment here?

File: src/main/java/org/lightcouch/View.java [hunk 6]
```diff
@@ -381,7 +383,7 @@ public boolean queryForBoolean() {
                 boolean isLastPage = resultRows <= rowsPerPage;
                 if (!isLastPage) {
                     page.setHasNext(true);
-                    jsonNext.addProperty(START_KEY, rows.get(i).getKey());
+                    jsonNext.addProperty(START_KEY, rows.get(i).getKey().toString());
```
Comment (mikerhodes) [inline src/main/java/org/lightcouch/View.java]: Why is `toString` required in this case and on line 377?

File: src/test/java/com/cloudant/tests/ViewsTest.java [hunk 7]
```diff
@@ -104,13 +105,238 @@ public void byStartAndEndKey() {
...
+     */
+
+    @Test
+    public void byStartAndEndBooleanKey() {
+        multiValueKeyInit();
+
+        View query = db.view("keyarray_example/boolean")
+                .startKey(false)
+                .endKey(false)
+                .includeDocs(true);
```
Comment (mikerhodes) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Do you need `includeDocs`, you don't look at the docs.

File: src/test/java/com/cloudant/tests/ViewsTest.java [hunk 8]
```diff
@@ -573,25 +799,138 @@ private Page checkPagesBackward(int currentPage, int numberOfPages, Page page, b
...
+
+        //Created single value map function for boolean key test
+        mapFunctions = new DesignDocument.MapReduce();
+
+        mapFunctions.setMap(
+                "function(doc) { emit([doc.boolean], null); }");
+
+        keyViews.put("boolean", mapFunctions);
+
+        designDoc.setViews(keyViews);
```
Comment (mikerhodes) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: You'd have much less code here if you just used a big JSON string for this ddoc -- does the library let you do that?

File: src/main/java/org/lightcouch/View.java (hunk 5 above)
Comment (emlaver) [inline src/main/java/org/lightcouch/View.java]: @mikerhodes I realized after this addition above that the following start/end keys are valid and can effect the pagination: [1], ["1"], ["abc"].  I added a note in FB that I was in progress of fixing this.  The idea would be to check if the key is an array containing one value, and then manipulating the value before setting the final page parameters for start/end key.

File: src/test/java/com/cloudant/tests/ViewsTest.java (hunk 7 above)
Comment (emlaver) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: You're right, the includeDocs are not needed.  They were just copied from another text case.

File: src/main/java/org/lightcouch/View.java (hunk 6 above)
Comment (emlaver) [inline src/main/java/org/lightcouch/View.java]: In order to fix the original problem with the queryView causing an exception, I changed the first parameter of the queryView constructor from String.class to Object.class (line 365).  Now, the `getKey()` method above will return an object.  I needed to call `toString` to get the string representation and set that to the start key.  

In my latest tests, I'm finding that key arrays with one string value are missing the quotes by the time it reaches this code.  For example, a key with "["abc"]" or "["1"]" would show as "[abc]" or "[1]".  Here is a gist of an if block that would add the required quotes:
https://gist.github.com/emlaver/81cc2db5cbfc0291ff8e

I'd like to replace the code in the gist with something more efficient.  I may need to manipulate the key in the `JsonToObject` method of CouchDBUtil.java.

File: src/test/java/com/cloudant/tests/ViewsTest.java (hunk 8 above)
Comment (emlaver) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: I was going to create map.js classes under the resources/example/views folder.  This is where the views for the rest of the test cases (like example/foo) exist.

**Comment by emlaver** [comment]
I've made several changes:
- Move the views that were originally hard-coded in the ViewsTest class to the design-docs/example directory
- Use JsonElement object instead of String for start/end key in View class. Some benefits of this change:
   - No need to manipulate the JSON within a String object where the syntax is not recognized
   - Easy to work with a JsonArray object for single or multi-value key arrays
- Add additional test cases (and more to come) for multi-value keys in query, queryPage, and with pagination.
- Mike's comment to remove `includeDocs`

**Comment by ricellis** [comment]
@emlaver is this PR still valid? I feel like it has been superseded by #71

**Comment by emlaver** [comment]
@ricellis Commit 2129358 contains fixes you commented on for the javadoc in ViewsTest class.

**Comment by emlaver** [comment]
@ricellis commit 15663de completely replaces ```getKeyAsJson``` with ```getKeyAsJsonElement```.  This also required the **key** class variable to be changed from a String to a JsonElement.

File: src/main/java/org/lightcouch/View.java [hunk 9]
```diff
@@ -428,7 +430,7 @@ public boolean queryForBoolean() {
      * @param key The key value, accepts a single value or multiple values for complex keys.
      */
     public View key(Object... key) {
-        this.key = getKeyAsJson(key);
+        this.key = getKeyAsJsonElement(key);
         uriBuilder.query("key", this.key);
```
Comment (mikerhodes) [inline src/main/java/org/lightcouch/View.java]: After the change to using JsonElement rather than String, to fix the bug with complex keys, you are implicitly relying on JsonElement's `toString` method returning a valid JSON string here for us as part of the URL, which isn't part of the `toString` method's contract. 

Use `Gson.toJson` to guarantee you are sending in valid JSON in the URL. 

The same goes for `startKey` and `endKey`, below.

File: src/main/java/org/lightcouch/View.java [hunk 10]
```diff
@@ -552,7 +554,7 @@ public View updateSeq(Boolean updateSeq) {
      * @return
      */
     public View keys(Object... keys) {
-        this.allDocsKeys = String.format("{%s:%s}", gson.toJson("keys"), getKeyAsJson(keys));
+        this.allDocsKeys = String.format("{%s:%s}", gson.toJson("keys"), getKeyAsJsonElement(keys));
```
Comment (mikerhodes) [inline src/main/java/org/lightcouch/View.java]: While you are here, fix this line to do something sensible:

```
JsonElement e = new JsonElement();
e.add("keys", getKeyAsJsonElement(keys));
gson.toJson(e);
```

File: src/test/java/com/cloudant/tests/ViewsTest.java [hunk 11]
```diff
@@ -107,13 +107,227 @@ public void byStartAndEndKey() {
...
+
+        page = null;
+        try {
+            page = query.queryPage(30, null, Object.class);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        assertThat(page.getResultList().size(), is(3));
+    }
```
Comment (mikerhodes) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: This is two tests, split it into two methods.

File: src/test/java/com/cloudant/tests/ViewsTest.java [hunk 12]
```diff
@@ -107,13 +107,227 @@ public void byStartAndEndKey() {
...
+
+        query = db.view("example/created")
+                .startKey(new Object[]{1})
+                .endKey(new Object[]{2000});
+
+
+        result = query.query(Object.class);
+
+        assertThat(result.size(), is(3));
+    }
```
Comment (mikerhodes) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Again, two tests, split up. Particularly because the method name isn't correct for the first test.

File: src/test/java/com/cloudant/tests/ViewsTest.java [hunk 13]
```diff
@@ -107,13 +107,227 @@ public void byStartAndEndKey() {
...
+
+        List<Object> result = query.query(Object.class);
+
+        assertThat(result.size(), is(3));
+
+        Page page = null;
+        try {
+            page = query.queryPage(30, null, Object.class);
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (mikerhodes) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Why isn't this a test failure?

File: src/test/java/com/cloudant/tests/ViewsTest.java [hunk 14]
```diff
@@ -107,13 +107,227 @@ public void byStartAndEndKey() {
...
+
+        page = null;
+        try {
+            page = query.queryPage(30, null, Object.class);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        assertThat(page.getResultList().size(), is(3));
+    }
```
Comment (mikerhodes) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Seems like there are three tests here?

File: src/test/java/com/cloudant/tests/ViewsTest.java [hunk 15]
```diff
@@ -107,13 +107,227 @@ public void byStartAndEndKey() {
...
+    public void byStartAndEndThreeKeyArray() {
+        init();
+
+        View query = db.view("example/boolean_creator_created")
+                .startKey(false, "uuid", 1)
+                .endKey(false, "uuid", 2000);
+
+        List<Object> result = query.query(Object.class);
+
+        assertThat(result.size(), is(3));
```
Comment (mikerhodes) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Why are we doing this in every test?

File: src/test/java/com/cloudant/tests/ViewsTest.java [hunk 16]
```diff
@@ -528,6 +923,63 @@ private void checkPagination(boolean descending, int docCount, int docsPerPage,
...
+     *
+     * @param descending  true if the view is in descending order, and false otherwise.
+     * @param docCount    the total number of documents in the view.
+     * @param docsPerPage the maximum number of documents per page in the view.
+     * @param useNewView  true for creating a new view for the query, and false otherwise.
+     * @param pageToPages the list of page numbers to page to.
+     */
+    private void checkPaginationWithMultiValueKey(boolean descending, boolean useNewView,
+                                                  int docCount, int docsPerPage,
+                                                  int... pageToPages) {
```
Comment (mikerhodes) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: This shouldn't be a vararg method, it makes things very hard to follow at call sites.

File: src/test/java/com/cloudant/tests/ViewsTest.java [hunk 17]
```diff
@@ -552,6 +1004,35 @@ private Page checkPagesForward(int currentPage, int numberOfPages, Page page, bo
...
+     * @param view          the {@link View} object to use to perform the query (or null to create
+     *                         a new
+     *                      view for the query).
+     * @param viewCount     the current view in the array of multi-value key views. This value is
+     *                      used when view parameter is null.
+     * @return the last page in the view.
+     */
+    private Page checkPagesForwardMultiValueKey(int currentPage, int numberOfPages, Page page, boolean
+            descending, int docCount, int docsPerPage, View view, int viewCount) {
+        for (int i = 0; i < numberOfPages; ++i) {
```
Comment (mikerhodes) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Why `++i` rather than the more common `i++`?

File: src/test/resources/design-docs/example/views/boolean/map.js [hunk 18]
```diff
@@ -0,0 +1,3 @@
+function(doc) {
+  emit(doc.contentArray[0].boolean, null);
+}
```
Comment (mikerhodes) [inline src/test/resources/design-docs/example/views/boolean/map.js]: Be sure all newly added files have trailing new lines. (Set up your editor to always add them).

File: src/test/java/com/cloudant/tests/ViewsTest.java [hunk 19]
```diff
@@ -267,6 +481,15 @@ public void paginationAscendingReusingViewAllTheWayInEachDirectionTwice() {
     }
 
     /**
+     * Run the same check as paginationAscendingReusingViewAllTheWayInEachDirectionTwice
+     * test on views using an array of multi-value keys.
+     */
+    @Test
+    public void paginationAscendingReusingMVKeyViewAllTheWayInEachDirectionTwice() {
+        checkPaginationWithMultiValueKey(false, 6, 2, 3, 1, 3, 1);
```
Comment (mikerhodes) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: You could use a simple builder approach using inner classes to make these tests much more readable.

```java
public class CheckPaginationWithMultiValueKey {
 public b() {
   return new CheckPaginationWithMultiValueKey();
  }
  public descending(bool descending) {
    this.descending = descending;
    return this;
  }
  /* etc for other args */
  public runTest() {
    /* do current content of checkPaginationWithMultiValueKey method here */
  }
}

**Comment by mikerhodes** [comment]
Looks like this does fix the bug, but introduces a couple of others that need fixing.

In addition, I'd like to see a number of style and readability fixes.

File: src/test/java/com/cloudant/tests/ViewsTest.java (hunk 11 above)
Comment (emlaver) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Fixed in e75a2f1

File: src/test/java/com/cloudant/tests/ViewsTest.java (hunk 12 above)
Comment (emlaver) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Fixed in e75a2f1

File: src/test/java/com/cloudant/tests/ViewsTest.java (hunk 13 above)
Comment (emlaver) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: This test case will pass but is not an array of integers for multi-value key testing.  Fixed in e75a2f1

File: src/test/java/com/cloudant/tests/ViewsTest.java (hunk 14 above)
Comment (emlaver) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Fixed in e75a2f1

File: src/test/java/com/cloudant/tests/ViewsTest.java (hunk 15 above)
Comment (emlaver) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: I've changed the start/end key to produce different results in all the multi-value key tests in e75a2f1.  What other test results should we check for?

File: src/test/java/com/cloudant/tests/ViewsTest.java (hunk 16 above)
Comment (emlaver) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: @mikerhodes What if I changed constructor to ```checkPaginationWithMultiValueKey(int docCount, int docsPerPage, boolean descending, boolean useNewView, int... pageToPages)``` to separate the single int parameters from the int vararg?

Comment (emlaver) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: @mikerhodes Never-mind the comment above.  I will go with your builder pattern approach.

File: src/test/java/com/cloudant/tests/ViewsTest.java (hunk 17 above)
Comment (emlaver) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: @mikerhodes this was the design in the original code.  I believe this was used because pages need to start at 1.

File: src/test/java/com/cloudant/tests/ViewsTest.java (hunk 16 above)
Comment (emlaver) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Fixed in 7078797

File: src/test/java/com/cloudant/tests/ViewsTest.java (hunk 19 above)
Comment (emlaver) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Fixed in 7078797

File: src/main/java/org/lightcouch/View.java (hunk 9 above)
Comment (emlaver) [inline src/main/java/org/lightcouch/View.java]: Fixed in 0db11c3

File: src/main/java/org/lightcouch/View.java (hunk 10 above)
Comment (emlaver) [inline src/main/java/org/lightcouch/View.java]: Fixed in 0db11c3

**Comment by ricellis** [comment]
+1

File: src/test/resources/design-docs/example/views/boolean/map.js (hunk 18 above)
Comment (emlaver) [inline src/test/resources/design-docs/example/views/boolean/map.js]: Fixed in 4de5a8d

**Comment by mikerhodes** [comment]
Looks great. +1. Looks like it can be squashed to just one commit before merging?

After spending the time reviewing the PR, I think the commit's title should be something like "Fix paginated view queries that use complex keys", and remember the `BugzID: ` bit.
